scalding-parquet-scrooge/src/main/scala/com/twitter/scalding/parquet/scrooge/ParquetCollectionFormatCompatibility.scala

```diff
@@ -0,0 +1,163 @@
+ * of collection group--list and map. This is currently used in [[ScroogeReadSupport]] where
+ * projected read schema can come from:
+ * 1) Thrift struct via [[org.apache.parquet.thrift.ThriftSchemaConvertVisitor]] which always
+ * describe list with `_tuple` format, and map which has `MAP_KEY_VALUE` annotation.
+ * 2) User-supplied schema string via config key
+ * [[org.apache.parquet.hadoop.api.ReadSupport.PARQUET_READ_SCHEMA]]
+ *
+ * By definition, the projected read schema is a "sub-graph" of file schema in terms of field names.
+ * (We do allow optional field in projected read schema to be in
+ * the projected file schema, even if file schema may not originally contain it.)
+ * The graphs of the two schemas may, however, differ for list and map type because of multiple
+ * legacy formats and the canonical one. This class supports all directions of conversion.
+ *
+ * The projection strategy is:
+ * 1) traverse the two schemas and maintain only the fields in the read schema.
+ * 2) find collection type indicated by `repeated` type, and delegate it to respective list/map formatter.
+ * 3) wrap back the formatted repeated type with group type from projected read schema. This
+ * means the optional/required remains the same as that from projected read schema.
+ */
+private[scrooge] object ParquetCollectionFormatCompatibility {
+
+  private val logger = LoggerFactory.getLogger(getClass)
+
+  /**
+   * Project file schema to contain the same fields as the given projected read schema.
+   * The result is projected file schema with the same optional/required fields as the
+   * projected read schema, but collection type format as the file schema.
+   *
+   * @param projectedReadSchema read schema specifying field projection
+   * @param fileSchema file schema to be projected
+   */
+  def projectFileSchema(fileSchema: MessageType, projectedReadSchema: MessageType): MessageType = {
+    val projectedFileSchema = projectFileType(fileSchema, projectedReadSchema, FieldContext()).asGroupType()
+    logger.debug(s"Projected read schema:\n${projectedReadSchema}\n" +
+      s"File schema:\n${fileSchema}\n" +
+      s"Projected file schema:\n${projectedFileSchema}")
+    new MessageType(projectedFileSchema.getName, projectedFileSchema.getFields)
+  }
+
+  /**
+   * Main recursion to get projected file type. Traverse given schemas, filter out unneeded
+   * fields, and format read schema's list/map node to file schema's structure.
+   * The formatting of repeated type is not to one-to-one node swapping because we also have to
+   * handle projection and possible nested collection types in the repeated type.
+   */
+  private def projectFileType(fileType: Type, projectedReadType: Type, fieldContext: FieldContext): Type = {
+    (extractCollectionGroup(projectedReadType), extractCollectionGroup(fileType)) match {
+      case _ if projectedReadType.isPrimitive && fileType.isPrimitive =>
+        projectedReadType
```

Comment:
Do we need to check that both fields are the _same_ primitive type? Or is that handled elsewhere? I imagine that certain types of primitive-type-changing projections might be supported (e.g. widening an int to a long) but others might not be (e.g. converting a string to an int).

Reply:
That type compatibility is handled in a separate place here, so by the time this is called we should already have type compatibility.
https://github.com/twitter/scalding/blob/73ce124f3c0156bd41f5394d2b609a5cd74f370e/scalding-parquet-scrooge/src/main/java/com/twitter/scalding/parquet/scrooge/ScroogeReadSupport.java#L133-L136